apps/microtvm/zephyr/template_project/microtvm_api_server.py

```diff
@@ -885,5 +945,186 @@ def _wait_for_qemu(self):
+        env["FVP_BIN_PATH"] = str(pathlib.Path(self.options["arm_fvp_path"]).parent)
+        env["ARMFVP_BIN_PATH"] = str(API_SERVER_DIR / "fvp-hack")
+        self.proc = subprocess.Popen(
+            args,
+            cwd=BUILD_DIR,
+            env=env,
+            stdout=subprocess.PIPE,
+        )
+        threading.Thread(target=self._fvp_check_stdout, daemon=True).start()
+
+        self.iris_port = self._wait_for_fvp()
+        _LOG.info("IRIS started on port %d", self.iris_port)
+        NetworkModelInitializer = self._iris_lib.NetworkModelInitializer.NetworkModelInitializer
+        self._model_init = NetworkModelInitializer(
+            host="localhost", port=self.iris_port, timeout_in_ms=1000
+        )
+        self._model = self._model_init.start()
+        self._target = self._model.get_target("component.FVP_MPS3_Corstone_SSE_300.cpu0")
+
+        self._target.handle_semihost_io()
+        self._target._stdout = BlockingStream()
+        self._target._stdin = BlockingStream()
+        self._model.run(blocking=False, timeout=100)
+        self._wait_for_semihost_init()
+
+        return server.TransportTimeouts(
+            session_start_retry_timeout_sec=2.0,
+            session_start_timeout_sec=10.0,
+            session_established_timeout_sec=10.0,
+        )
+
+    def _fvp_check_stdout(self):
+        for line in self.proc.stdout:
+            line = str(line, "utf-8")
+            m = re.match(r"Iris server started listening to port ([0-9]+)\n", line)
+            n = re.match("microTVM Zephyr runtime - running", line)
+            if m:
+                self._queue.put((ZephyrFvpMakeResult.FVP_STARTED, int(m.group(1))))
+            elif n:
+                self._queue.put((ZephyrFvpMakeResult.FVP_INIT, None))
+            else:
+                line = re.sub("[^a-zA-Z0-9 \n]", "", line)
+                pattern = r"recipe for target (\w*) failed"
+                if re.search(pattern, line, re.IGNORECASE):
+                    self._queue.put((ZephyrFvpMakeResult.MAKE_FAILED, None))
+
+        self._queue.put((ZephyrFvpMakeResult.EOF, None))
+
+    def _wait_for_fvp(self):
+        while True:
+            try:
+                item = self._queue.get(timeout=120)
+            except Exception:
+                raise TimeoutError("FVP setup timeout.")
+
+            if item[0] == ZephyrFvpMakeResult.FVP_STARTED:
+                return item[1]
+
+            if item[0] in [ZephyrFvpMakeResult.MAKE_FAILED, ZephyrFvpMakeResult.EOF]:
+                raise RuntimeError("FVP setup failed.")
+
+            raise ValueError(f"{item} not expected.")
+
+    def _wait_for_semihost_init(self):
+        while True:
+            try:
+                item = self._queue.get(timeout=120)
+            except Exception:
+                raise TimeoutError("semihost init timeout.")
+
+            if item[0] == ZephyrFvpMakeResult.FVP_INIT:
+                return
+
+            raise ValueError(f"{item} not expected.")
```

Comment:
could you add some context to these two errors? i.e. what is the API server doing right now and what does it expect?

Reply:
The API is waiting for a specific line to appear on the stdout. I added a comment to make it more clear.